### ISSUE TITLE
Support for Sparse Algebras 

### ISSUE THREAD

**Issue by Stefan-Endres** [pygae/clifford#40]
In certain high-dimensional applications we only make use of a subspace of grades. For example the Cl(n) Clifford algebra grows in dimensionality by `O(2^n)` while using only blades of grade `k = 1` and `k = n - 1` grows linearly by `O(2n)`.

I wrote a quick naive implementation just to test initiation times, the `kr` argument is used to specify a list of desired grades.

https://github.com/Stefan-Endres/clifford/commit/81c78c3cd36b3d952ec652072167af786d874116

```
>>> cf.Cl(4, kr=[1, 3])
(Layout([1, 1, 1, 1], [(), (1,), (2,), (3,), (4,), (1, 2, 3), (1, 2, 4), (1, 3, 4), (2, 3, 4)], 
firstIdx=1, 
names=['', 'e1', 'e2', 'e3', 'e4', 'e123', 'e124', 'e134', 'e234']), 
{'e1': (1^e1), 'e2': (1^e2), 'e3': (1^e3), 'e4': (1^e4), 
'e123': (1^e123), 'e124': (1^e124), 'e134': (1^e134), 'e234': 
(1^e234)})

>>> cf.Cl(4)
(Layout([1, 1, 1, 1], [(), (1,), (2,), (3,), (4,), (1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4), (1, 2, 3), (1, 2, 4), (1, 3, 4), (2, 3, 4), (1, 2, 3, 4)], 
firstIdx=1, 
names=['', 'e1', 'e2', 'e3', 'e4', 'e12', 'e13', 'e14', 'e23', 'e24', 'e34', 'e123', 'e124', 'e134', 'e234', 'e1234']), 
{'e1': (1^e1), 'e2': (1^e2), 'e3': (1^e3), 'e4': (1^e4), 
'e12': (1^e12), 'e13': (1^e13), 'e14': (1^e14), 'e23': (1^e23), 'e24': (1^e24), 'e34': (1^e34),
'e123': (1^e123), 'e124': (1^e124), 'e134': (1^e134), 'e234': (1^e234), 'e1234':
(1^e1234)})
```
Some basic operations such as mutlivector addition, grade projection etc. are preserved nicely. Others such as dual (in cases where the result is an initiated grade) and reflection are not.

The initiation time growth isn't as promising as I'd hoped for either, but could possibly be drastically improved with a less naive implementation. In particular 99.1% of the initiation is spent in this loop in  `_getEvenOdd`: https://github.com/pygae/clifford/blob/8793d3e4faf758770ae80d8d9e111574677772bc/clifford/__init__.py#L567

There is a comment in the code that states the loop runs for `grade!` permutations, however, at the moment it is unclear to me how the rest of the algebra and code will be affected when I change the range of these permutations or if it is even still looping for all the canonical grades at the moment. Especially when supplied with more arbitrary grade ranges.

Initiation times:

```
cf.Cl(4)
Initiation time: 0.22497177124023438
cf.Cl(4, kr=[1, 3]) 
Initiation time: 0.0009617805480957031
cf.Cl(5)
Initiation time: 0.009763717651367188
cf.Cl(5, kr=[1, 4]) 
Initiation time: 0.0020651817321777344
cf.Cl(6)
Initiation time: 0.0555272102355957
cf.Cl(6, kr=[1, 5]) 
Initiation time: 0.010433673858642578
cf.Cl(7)
Initiation time: 0.41026782989501953
cf.Cl(7, kr=[1, 6]) 
Initiation time: 0.08707642555236816
cf.Cl(8)
Initiation time: 4.242865324020386
cf.Cl(8, kr=[1, 7]) 
Initiation time: 1.0086712837219238
cf.Cl(9)
Initiation time: 61.97481608390808
cf.Cl(9, kr=[1, 8]) 
Initiation time: 15.875498056411743
# cf.Cl(10) did not initiate after several minutes 
f.Cl(10, kr=[1, 9]) 
Initiation time: 329.20880341529846

```
cProfile for `cf.Cl(9, kr=[1, 8]`:

Name | Call Count | Time (ms) | Own Time (ms)
-- | -- | -- | --
__init__ | 1 | 23157 | 0
Cl | 1 | 23157 | 0
_genEvenOdd | 1 | 23155 | 973
modify_idx | 362880 | 20229 | 11265
typeof | 3265940 | 8964 | 1721
wrapper | 3265942 | 6188 | 1311
_typeof_int | 3265937 | 3374 | 1064
bit_length | 3265937 | 2310 | 1436

I would be happy to contribute as much as possible under the direction of the developers.

Regards, 
Stefan Endres

**Comment by hugohadfield** [comment]
Thanks for the detailed issue @Stefan-Endres ! High dimensional algebra support is a big priority for us at the moment as it is the biggest performance killer in this library at the moment and people are interested in it, it would be great to get your contributions.
I have been thinking about our problems here for a little while, check out https://github.com/pygae/clifford/issues/17
I'm just digging into this even odd issue now and it is pretty nasty

**Comment by hugohadfield** [comment]
These even and odd permutation tables appear to be used only in _gmtElement. I think modifying _genEvenOdd to use your kr parameter and limiting the  for i in range(self.gaDims): to the indices of your sparse algebra will work fine and will bring big speedups

**Comment by Stefan-Endres** [comment]
Wow! Thank you, it's already much faster. I'm going to play around with weaving kr into the _genEvenOdd loop hopefully we can squeeze even more speed out of a sparse initialization.

**Comment by moble** [comment]
I'd just like to suggest that since `Cl` is pretty much the main callable in this library, it might be helpful to have a more descriptive name for the new argument.  Maybe `desired_grades` or even just `grades`?

**Comment by enkimute** [comment]
Just for comparison, ganja.js absolutely not optimized generation times :

```
1-dimensional clifford: 0.674072265625ms
2-dimensional clifford: 0.808837890625ms
3-dimensional clifford: 2.743896484375ms
4-dimensional clifford: 10.283935546875ms
5-dimensional clifford: 22.559814453125ms
6-dimensional clifford: 83.708984375ms
7-dimensional clifford: 439.293212890625ms
8-dimensional clifford: 2795.296875ms
9-dimensional clifford: 18897.85400390625ms
```

**Comment by hugohadfield** [comment]
Thanks for the table @enkimute interesting how fast the javascript startup times are!
just added to my PR here: https://github.com/pygae/clifford/pull/43
this should bring the start up times down very significantly

**Comment by enkimute** [comment]
After fixing some very obvious things :

```
1-dimensional clifford: 2.13525390625ms
2-dimensional clifford: 1.583740234375ms
3-dimensional clifford: 1.738037109375ms
4-dimensional clifford: 7.637939453125ms
5-dimensional clifford: 16.21484375ms
6-dimensional clifford: 45.40087890625ms
7-dimensional clifford: 195.120849609375ms
8-dimensional clifford: 954.40087890625ms
9-dimensional clifford: 5272.82421875ms
```

**Comment by hugohadfield** [comment]
@Stefan-Endres with the latest release of the library we are in a place where higher dimensional algebra initialisation is pretty fast, do you think sparse algebra support is still necessary? Or should we close the issue?

**Comment by Stefan-Endres** [comment]
I think it would be better to close the issue for now. The last time I worked on this I was not able to make all the desirable operations work seamlessly, what I did relies on some very ugly/unmaintanable hacks. If I find a better solution some day I'll open a new PR if there are applications requiring very high dimensions (meaning to the point where the memory requirements of combinatorial growth of the blades of "middle" grades are intractable) and/or a good maintainable solution, but for now I don't think it would benefit the library (on a personal note it's working really well for my applications at the moment as I do not use extremely high dimensions).

Thanks for the great work on the initialisation improvements!
